typing/typecore.ml

```diff
@@ -981,102 +981,67 @@ type pattern_checking_mode =
   (** We are checking a counter-example to exhaustivity
       generated by the pattern-matching analyser. *)
 
-
 (** In [Counter_example] mode the responsibility of [type_pat] is to
-   check that a counter-example to exhaustivity is valid: it has to be
-   type-correct in a slightly different sense than for user-code, see
-   [splitting_mode] below, and also capture at least one concrete
-   value.
+   check that a counter-example to exhaustivity is valid: whether there is a
+   choice of clauses in or-patterns (including explosions of wild cards at
+   some type), which we can view as a path in a search tree, that leads
+   to a valid pattern.
+   Invalid paths are typically due to incompatible GADT equations,
+   or empty sum types.
 
    [counter_example_checking_info] contains various parameters for
    this mode of operation. *)
 and counter_example_checking_info = {
-    explode: explosion_fuel;
+    explosion_fuel: int;
     splitting_mode: splitting_mode;
-    synthetic_names: counter_example_synthetic_names;
+    constrs: (string, Types.constructor_description) Hashtbl.t;
+    labels: (string, Types.label_description) Hashtbl.t;
   }
+(** [explosion_fuel] is the depth limit on wild card explosion.
+    It is required to avoid non-termination and compilation-time slowdowns.
+    Explosion converts a wildcard to an or-pattern of the constructors of its
+    type for a variant type, or a tuple of wildcards for unary constructors,
+    records, or tuples.
+
+    [splitting_mode] controls the handling of or-patterns.
+    When typing an or-pattern in [Counter_example] mode, we only need to
+    select one branch that leads to a valid path.
+    [Backtrack_or] does this by using backtracking, trying to choose a
+    branch in turn, and to complete it into a valid path.
+    [Refine_or] tries another approach.
+    It first attempts to extract the branches that do not introduce new
+    constraints. Those branches are such that, if they fail, all other
+    branches will fail.
+    If successful, it attempts to complete the path, ignoring other branches.
+    This avoids unnecessary backtracking.
+    If all branches are constrained, it falls back to backtracking.
+
+    [constrs] and [labels] map names generated by [Parmatch] to internal
+    description. [Parmatch] produces counter-examples by turning
+    typed patterns into [Parsetree.pattern]. In this process, constructor
+    and label paths are lost, and have to be replace by generated strings,
```

Comment:
"have to be replaced*".

Also, they do not have to, but they are.

Reply:
done